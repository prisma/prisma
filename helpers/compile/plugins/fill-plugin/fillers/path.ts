export function resolve(...args: any[]) {
  return args.join('/')
}

export function join(...args: any[]) {
  return args.join('/')
}

export const sep = '/'

export const posix = {
  sep,
}

/**
 * A poor man's shim for the "path" module
 */
const path = {
  resolve,
  posix,
  join,
  sep,
}

export default path
