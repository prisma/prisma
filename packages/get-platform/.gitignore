node_modules
.DS_Store
query-engine*
yarn-error.log
pnpm-debug.log

*.db

dist/

copy-photon.sh
*.tsbuildinfo