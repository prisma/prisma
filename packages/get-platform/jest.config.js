module.exports = {
  transform: {
    '^.+\\.(m?j|t)s$': '@swc/jest',
  },
  transformIgnorePatterns: [],
  testEnvironment: 'node',
  collectCoverage: process.env.CI ? true : false,
  coverageReporters: ['clover'],
  coverageDirectory: 'src/__tests__/coverage',
  collectCoverageFrom: ['src/**/*.ts', '!**/__tests__/**/*'],
  testMatch: ['**/src/__tests__/**/*.test.ts'],
}
