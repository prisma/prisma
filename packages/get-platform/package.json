{
  "name": "@prisma/get-platform",
  "version": "0.0.0",
  "description": "This package is intended for Prisma's internal use",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "license": "Apache-2.0",
  "author": "Tim Suchanek <suchanek@prisma.io>",
  "homepage": "https://www.prisma.io",
  "repository": {
    "type": "git",
    "url": "https://github.com/prisma/prisma.git",
    "directory": "packages/get-platform"
  },
  "bugs": "https://github.com/prisma/prisma/issues",
  "devDependencies": {
    "@swc/core": "1.3.14",
    "@swc/jest": "0.2.24",
    "@types/jest": "29.2.4",
    "@types/node": "16.18.11",
    "jest": "29.3.1",
    "typescript": "4.8.4"
  },
  "dependencies": {
    "@prisma/debug": "workspace:*",
    "ts-pattern": "4.0.6"
  },
  "scripts": {
    "dev": "DEV=true node -r esbuild-register helpers/build.ts",
    "build": "node -r esbuild-register helpers/build.ts",
    "test": "jest",
    "prepublishOnly": "pnpm run build"
  },
  "files": [
    "README.md",
    "dist"
  ],
  "sideEffects": false
}
