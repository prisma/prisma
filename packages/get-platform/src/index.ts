export { getNodeAPIName } from './getNodeAPIName'
export { getos, getPlatform } from './getPlatform'
export { isNodeAPISupported } from './isNodeAPISupported'
export { type Platform, platforms } from './platforms'
