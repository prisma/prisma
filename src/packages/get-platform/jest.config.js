module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node',
  collectCoverage: true,
  coverageDirectory: 'src/__tests__/coverage',
  collectCoverageFrom: ['src/**/*.ts', '!**/__tests__/**/*'],
  testMatch: ['**/src/__tests__/**/*.test.ts'],
  globals: {
    'ts-jest': {
      packageJson: 'package.json',
    },
  },
}
