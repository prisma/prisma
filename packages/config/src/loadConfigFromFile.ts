import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'

import { Debug } from '@prisma/driver-adapter-utils'
import { Either, pipe } from 'effect'
import { ParseError } from 'effect/ParseResult'

import { defineConfig, type PrismaConfigInternal } from './defineConfig'
import { parsePrismaConfigInternalShape, parsePrismaConfigShape } from './PrismaConfig'

const debug = Debug('prisma:config:loadConfigFromFile')

type LoadConfigFromFileInput = {
  /**
   * The path to the config file to load. If not provided, we will attempt to find a config file in the `configRoot` directory.
   */
  configFile?: string

  /**
   * The directory to search for the config file in. Defaults to the current working directory.
   */
  configRoot?: string
}

export type LoadConfigFromFileError =
  | {
      _tag: 'ConfigFileNotFound'
    }
  | {
      _tag: 'TypeScriptImportFailed'
      error: Error
    }
  | {
      _tag: 'ConfigFileParseError'
      error: ParseError
    }
  | {
      _tag: 'UnknownError'
      error: Error
    }

export type ConfigFromFile =
  | {
      resolvedPath: string
      config: PrismaConfigInternal
      error?: never
    }
  | {
      resolvedPath: string
      config?: never
      error: LoadConfigFromFileError
    }
  | {
      resolvedPath: null
      config?: never
      error?: never
    }

/**
 * Load a Prisma config file from the given directory.
 * This function may fail, but it will never throw.
 * The possible error is returned in the result object, so the caller can handle it as needed.
 */

export async function loadConfigFromFile({
  configFile,
  configRoot = process.cwd(),
}: LoadConfigFromFileInput): Promise<ConfigFromFile> {
  const start = performance.now()
  const getTime = () => `${(performance.now() - start).toFixed(2)}ms`

  let resolvedPath: string | null

  if (configFile) {
    // The user gave us a specific config file to load. If it doesn't exist, we should raise an error.
    resolvedPath = path.resolve(configRoot, configFile)

    if (!fs.existsSync(resolvedPath)) {
      debug(`The given config file was not found at %s`, resolvedPath)
      return { resolvedPath, error: { _tag: 'ConfigFileNotFound' } }
    }
  } else {
    // We attempt to find a config file in the given directory. If none is found, we should return early, without errors.
    resolvedPath =
      ['prisma.config.ts'].map((file) => path.resolve(configRoot, file)).find((file) => fs.existsSync(file)) ?? null

    if (resolvedPath === null) {
      debug(`No config file found in the current working directory %s`, configRoot)

      return { resolvedPath }
    }
  }

  try {
    const { required, error } = await requireTypeScriptFile(resolvedPath)

    if (error) {
      return {
        resolvedPath,
        error,
      }
    }

    debug(`Config file loaded in %s`, getTime())

    const defaultExport = required['default']

    const parseResultEither = pipe(
      // If the given config conforms to the `PrismaConfig` shape, feed it to `defineConfig`.
      parsePrismaConfigShape(defaultExport),
      Either.map((config) => {
        debug('Parsed `PrismaConfig` shape: %o', config)
        return defineConfig(config)
      }),
      // Otherwise, try to parse it as a `PrismaConfigInternal` shape.
      Either.orElse(() => parsePrismaConfigInternalShape(defaultExport)),
    )

    // Failure case
    if (Either.isLeft(parseResultEither)) {
      return {
        resolvedPath,
        error: {
          _tag: 'ConfigFileParseError',
          error: parseResultEither.left,
        },
      }
    }

    process.stdout.write(`Loaded Prisma config from "${resolvedPath}".\n`)

    // Success case
    const prismaConfig = transformPathsInConfigToAbsolute(parseResultEither.right, resolvedPath)

    return {
      config: {
        ...prismaConfig,
        loadedFromFile: resolvedPath,
      },
      resolvedPath,
    }
  } catch (e) {
    // As far as we know, this branch should be unreachable.
    const error = e as Error

    return {
      resolvedPath,
      error: {
        _tag: 'UnknownError',
        error,
      },
    }
  }
}

// Note: `esbuild-register` combines well with `esbuild`, which we already use.
// However, we might consider adopting `jiti` in the future, either directly or
// via `c12`.
async function requireTypeScriptFile(resolvedPath: string) {
  try {
    const { register: esbuildRegister } = await import('esbuild-register/dist/node')

    // eslint-disable-next-line @typescript-eslint/unbound-method
    const { unregister } = esbuildRegister({
      format: 'cjs',
      loader: 'ts',
    })
    const configExport = require(resolvedPath)
    unregister()

    return {
      required: configExport,
      error: null,
    }
  } catch (e) {
    const error = e as Error
    debug('esbuild-register registration failed: %s', error.message)

    return {
      error: {
        _tag: 'TypeScriptImportFailed',
        error,
      } as const,
    }
  }
}

function transformPathsInConfigToAbsolute(
  prismaConfig: PrismaConfigInternal,
  resolvedPath: string,
): PrismaConfigInternal {
  if (prismaConfig.schema?.kind === 'single') {
    return {
      ...prismaConfig,
      schema: {
        ...prismaConfig.schema,
        filePath: path.resolve(path.dirname(resolvedPath), prismaConfig.schema.filePath),
      },
    }
  } else if (prismaConfig.schema?.kind === 'multi') {
    return {
      ...prismaConfig,
      schema: {
        ...prismaConfig.schema,
        folderPath: path.resolve(path.dirname(resolvedPath), prismaConfig.schema.folderPath),
      },
    }
  } else {
    return prismaConfig
  }
}
