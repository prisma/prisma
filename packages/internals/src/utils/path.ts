import path from 'path'

/**
 * Normalize `filePath` to use forward slashes as a separator. `filePath` is
 * treated as a path specific to the current platform, so backslashes will only
 * be replaced with forward slashes on Windows. On other platforms, where a
 * backslash is a valid filename character, it will be treated as such and will
 * not be replaced.
 */
export function pathToPosix(filePath: string): string {
  if (path.sep === path.posix.sep) {
    return filePath
  }
  return filePath.split(path.sep).join(path.posix.sep)
}

/**
 * Returns the longest common path ancestor of the two paths (which may also be equal to one or both of them).
 * If they don't share a common ancestor (which may be the case on Windows if both paths are on different disks),
 * returns `undefined`.
 */
export function longestCommonPathPrefix(pathA: string, pathB: string): string | undefined {
  if (!path.isAbsolute(pathA) || !path.isAbsolute(pathB)) {
    throw new Error('longestCommonPathPrefix expects absolute paths')
  }

  if (process.platform === 'win32' && (pathA.startsWith('\\\\') || pathB.startsWith('\\\\'))) {
    // Make both paths namespaced if at least one of them is.
    pathA = path.toNamespacedPath(pathA)
    pathB = path.toNamespacedPath(pathB)
  }

  const commonPrefix = longestCommonPrefix(pathA.split(path.sep), pathB.split(path.sep)).join(path.sep)

  if (commonPrefix === '') {
    return process.platform === 'win32' ? undefined : '/'
  }

  if (process.platform === 'win32' && ['\\', '\\\\?', '\\\\.'].includes(commonPrefix)) {
    return undefined
  }

  if (process.platform === 'win32' && commonPrefix.endsWith(':')) {
    // Disk specifier without a backslash at the end is not an absolute path on windows,
    // it refers to current working directory on that disk.
    return commonPrefix + '\\'
  }

  return commonPrefix
}

function longestCommonPrefix<T>(sequenceA: T[], sequenceB: T[]): T[] {
  const maxLen = Math.min(sequenceA.length, sequenceB.length)
  let sliceLen = 0

  while (sliceLen <= maxLen && sequenceA[sliceLen] === sequenceB[sliceLen]) {
    sliceLen++
  }

  return sequenceA.slice(0, sliceLen)
}
