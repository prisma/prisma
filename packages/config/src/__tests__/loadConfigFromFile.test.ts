import path from 'node:path'

import { jestContext } from '@prisma/get-platform'
import type { ParseError } from 'effect/ParseResult'

import { loadConfigFromFile, type LoadConfigFromFileError } from '../loadConfigFromFile'

const ctx = jestContext.new().assemble()

describe('loadConfigFromFile', () => {
  function assertErrorTypeScriptImportFailed(error: LoadConfigFromFileError | undefined): asserts error is {
    _tag: 'TypeScriptImportFailed'
    error: Error
  } {
    expect(error).toMatchObject({ _tag: 'TypeScriptImportFailed' })
  }

  function assertErrorConfigFileParseError(error: LoadConfigFromFileError | undefined): asserts error is {
    _tag: 'ConfigFileParseError'
    error: ParseError
  } {
    expect(error).toMatchObject({ _tag: 'ConfigFileParseError' })
  }

  describe('no-define-config', () => {
    it('successfully loads a Prisma config file that does not use the `defineConfig` function', async () => {
      ctx.fixture('loadConfigFromFile/no-define-config')
      const cwd = ctx.fs.cwd()

      const { config, error, resolvedPath } = await loadConfigFromFile({})
      expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), 'prisma.config.ts'))
      expect(error).toBeUndefined()
      expect(config).toMatchObject({
        earlyAccess: true,
        schema: {
          kind: 'single',
          filePath: path.join(cwd, 'schema.prisma'),
        },
      })
    })
  })

  describe('schema', () => {
    describe('single', () => {
      it('succeeds when it points to a single Prisma schema file that exists via an absolute path', async () => {
        ctx.fixture('loadConfigFromFile/schema/single-exists')
        const cwd = ctx.fs.cwd()

        const { config, error, resolvedPath } = await loadConfigFromFile({})
        expect(resolvedPath).toMatch(path.join(cwd, 'prisma.config.ts'))
        expect(error).toBeUndefined()
        expect(config).toMatchObject({
          earlyAccess: true,
          loadedFromFile: resolvedPath,
          schema: {
            kind: 'single',
            filePath: path.join(cwd, 'prisma', 'schema.prisma'),
          },
        })
      })

      it('succeeds when it points to a single Prisma schema file that exists via a relative path', async () => {
        ctx.fixture('loadConfigFromFile/schema/single-exists-relative')
        const cwd = ctx.fs.cwd()

        const { config, error, resolvedPath } = await loadConfigFromFile({})
        expect(resolvedPath).toMatch(path.join(cwd, 'prisma.config.ts'))
        expect(error).toBeUndefined()
        expect(config).toMatchObject({
          earlyAccess: true,
          loadedFromFile: resolvedPath,
          schema: {
            kind: 'single',
            filePath: path.join(cwd, 'prisma', 'schema.prisma'),
          },
        })
      })

      it('succeeds when it points to a single Prisma schema file that does not exists', async () => {
        ctx.fixture('loadConfigFromFile/schema/single-does-not-exist')
        const cwd = ctx.fs.cwd()

        const { config, error, resolvedPath } = await loadConfigFromFile({})
        expect(resolvedPath).toMatch(path.join(cwd, 'prisma.config.ts'))
        expect(error).toBeUndefined()
        expect(config).toMatchObject({
          earlyAccess: true,
          loadedFromFile: resolvedPath,
          schema: {
            kind: 'single',
            filePath: path.join(cwd, 'prisma', 'schema.prisma'),
          },
        })
      })
    })

    describe('multi', () => {
      it('succeeds when it points to multiple Prisma schema files that exist via an absolute path', async () => {
        ctx.fixture('loadConfigFromFile/schema/multi-exist')
        const cwd = ctx.fs.cwd()

        const { config, error, resolvedPath } = await loadConfigFromFile({})
        expect(resolvedPath).toMatch(path.join(cwd, 'prisma.config.ts'))
        expect(error).toBeUndefined()
        expect(config).toMatchObject({
          earlyAccess: true,
          loadedFromFile: resolvedPath,
          schema: {
            kind: 'multi',
            folderPath: path.join(cwd, 'prisma', 'schema'),
          },
        })
      })

      it('succeeds when it points to multiple Prisma schema files that exist via a relative path ', async () => {
        ctx.fixture('loadConfigFromFile/schema/multi-exist-relative')
        const cwd = ctx.fs.cwd()

        const { config, error, resolvedPath } = await loadConfigFromFile({})
        expect(resolvedPath).toMatch(path.join(cwd, 'prisma.config.ts'))
        expect(error).toBeUndefined()
        expect(config).toMatchObject({
          earlyAccess: true,
          loadedFromFile: resolvedPath,
          schema: {
            kind: 'multi',
            folderPath: path.join(cwd, 'prisma', 'schema'),
          },
        })
      })

      it('succeeds when it points to multiple Prisma schema files that do not exist', async () => {
        ctx.fixture('loadConfigFromFile/schema/multi-do-not-exist')
        const cwd = ctx.fs.cwd()

        const { config, error, resolvedPath } = await loadConfigFromFile({})
        expect(resolvedPath).toMatch(path.join(cwd, 'prisma.config.ts'))
        expect(error).toBeUndefined()
        expect(config).toMatchObject({
          earlyAccess: true,
          loadedFromFile: resolvedPath,
          schema: {
            kind: 'multi',
            folderPath: path.join(cwd, 'prisma', 'schema'),
          },
        })
      })
    })
  })

  describe('invalid', () => {
    it('fails with `TypeScriptImportFailed` when the Prisma config file has a syntax error', async () => {
      ctx.fixture('loadConfigFromFile/invalid/syntax-error')

      const { config, error, resolvedPath } = await loadConfigFromFile({})
      expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), 'prisma.config.ts'))
      expect(config).toBeUndefined()
      assertErrorTypeScriptImportFailed(error)
      expect(error.error.message.replaceAll(resolvedPath!, '<prisma-config>.ts')).toMatchInlineSnapshot(`
        "  [31mÃ—[0m Unexpected eof
           â•­â”€[[36;1;4m<prisma-config>.ts[0m:5:3]
         [2m3[0m â”‚ export default defineConfig({
         [2m4[0m â”‚   earlyAccess: true,
         [2m5[0m â”‚ }
           â•°â”€â”€â”€â”€


        Caused by:
            Syntax Error"
      `)
    })

    it('fails with `ConfigFileParseError` when the Prisma config file has no default export', async () => {
      ctx.fixture('loadConfigFromFile/invalid/no-default-export')

      const { config, error, resolvedPath } = await loadConfigFromFile({})
      expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), 'prisma.config.ts'))
      expect(config).toBeUndefined()
      assertErrorConfigFileParseError(error)
      expect(error.error.message.replaceAll(resolvedPath!, '<prisma-config>.ts')).toMatchInlineSnapshot(
        `"Expected { readonly earlyAccess: true; readonly schema?: { readonly kind: "single"; readonly filePath: string } | { readonly kind: "multi"; readonly folderPath: string } | undefined; readonly loadedFromFile: string | null }, actual undefined"`,
      )
    })

    it(`fails with \`ConfigFileParseError\` when the default export in the Prisma config file does
        not conform to the expected schema shape`, async () => {
      ctx.fixture('loadConfigFromFile/invalid/no-schema-shape-conformance')

      const { config, error, resolvedPath } = await loadConfigFromFile({})
      expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), 'prisma.config.ts'))
      expect(config).toBeUndefined()
      assertErrorConfigFileParseError(error)
      expect(error.error.message.replaceAll(resolvedPath!, '<prisma-config>.ts')).toMatchInlineSnapshot(`
        "{ readonly earlyAccess: true; readonly schema?: { readonly kind: "single"; readonly filePath: string } | { readonly kind: "multi"; readonly folderPath: string } | undefined; readonly loadedFromFile: string | null }
        â””â”€ ["thisShouldFail"]
           â””â”€ is unexpected, expected: "earlyAccess" | "schema" | "loadedFromFile""
      `)
    })
  })

  describe('default-location', () => {
    it('succeeds when the Prisma config file exists and is in a valid format', async () => {
      ctx.fixture('loadConfigFromFile/default-location/success')

      const { config, error, resolvedPath } = await loadConfigFromFile({})
      expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), 'prisma.config.ts'))
      expect(config).toMatchObject({
        earlyAccess: true,
        loadedFromFile: resolvedPath,
      })
      expect(error).toBeUndefined()
    })

    it('returns no config when the Prisma config file does not exist', async () => {
      ctx.fixture('loadConfigFromFile/default-location/ignore')

      const { config, error, resolvedPath } = await loadConfigFromFile({})
      expect(resolvedPath).toBeNull()
      expect(error).toBeUndefined()
      expect(config).toBeUndefined()
    })
  })

  describe('custom-location', () => {
    beforeEach(() => {
      ctx.fixture('loadConfigFromFile/custom-location')
    })

    it('fails when the given file does not exist', async () => {
      const customConfigPath = path.join('does', 'not', 'exist.ts')

      const { config, error, resolvedPath } = await loadConfigFromFile({
        configFile: customConfigPath,
      })
      expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), customConfigPath))
      expect(error).toMatchObject({ _tag: 'ConfigFileNotFound' })
      expect(config).toBeUndefined()
    })

    it('succeeds when TypeScript file exists and is in a valid format', async () => {
      const customConfigPath = path.join('valid', 'prisma-config.ts')

      const { config, error, resolvedPath } = await loadConfigFromFile({
        configFile: customConfigPath,
      })
      expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), customConfigPath))
      expect(error).toBeUndefined()
      expect(config).toMatchObject({
        earlyAccess: true,
        loadedFromFile: resolvedPath,
      })
    })
  })

  it('typescript-cjs-ext-ts', async () => {
    ctx.fixture('loadConfigFromFile/typescript-cjs-ext-ts')

    const { config, error, resolvedPath } = await loadConfigFromFile({})
    expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), 'prisma.config.ts'))
    expect(config).toMatchObject({
      earlyAccess: true,
      loadedFromFile: resolvedPath,
    })
    expect(error).toBeUndefined()
  })

  it('typescript-esm-ext-ts', async () => {
    ctx.fixture('loadConfigFromFile/typescript-esm-ext-ts')

    const { config, error, resolvedPath } = await loadConfigFromFile({})
    expect(resolvedPath).toMatch(path.join(ctx.fs.cwd(), 'prisma.config.ts'))
    expect(config).toMatchObject({
      earlyAccess: true,
      loadedFromFile: resolvedPath,
    })
    expect(error).toBeUndefined()
  })

  describe('environment variables', () => {
    let processEnvBackup: NodeJS.ProcessEnv

    beforeEach(() => {
      processEnvBackup = { ...process.env }
    })

    afterEach(() => {
      process.env = processEnvBackup
    })

    function assertLoadConfigFromFileErrorIsUndefined(
      error: LoadConfigFromFileError | undefined,
    ): asserts error is undefined {
      expect(error).toBeUndefined()
    }

    test('if no custom env-var loading function is imported, it should skip loading any environment variables', async () => {
      ctx.fixture('loadConfigFromFile/env-baseline')
      const { config, error } = await loadConfigFromFile({})
      assertLoadConfigFromFileErrorIsUndefined(error)
      expect(config).toMatchObject({
        earlyAccess: true,
      })

      expect(process.env).toMatchObject(processEnvBackup)
      expect(process.env.TEST_CONNECTION_STRING).toBeUndefined()
    })

    test('if a sync custom env-var loading function is imported, it should load environment variables using the provided function', async () => {
      ctx.fixture('loadConfigFromFile/env-load-cjs')
      const { config, error } = await loadConfigFromFile({})
      assertLoadConfigFromFileErrorIsUndefined(error)
      expect(config).toMatchObject({
        earlyAccess: true,
      })

      expect(process.env).toMatchObject({
        ...processEnvBackup,
        TEST_CONNECTION_STRING: 'postgres://test-connection-string-from-env',
      })
    })

    test('if an async custom env-var loading function is used, it should fail loading environment variables using the provided function', async () => {
      ctx.fixture('loadConfigFromFile/env-load-esm')
      const { config, error } = await loadConfigFromFile({})

      expect(config).toBeUndefined()
      assertErrorTypeScriptImportFailed(error)
      expect(error).toMatchObject({ _tag: 'TypeScriptImportFailed' })
      expect(error.error).toMatchInlineSnapshot(
        `[SyntaxError: await is only valid in async functions and the top level bodies of modules]`,
      )

      expect(process.env).toMatchObject(processEnvBackup)
      expect(process.env.TEST_CONNECTION_STRING).toBeUndefined()
    })
  })
})
