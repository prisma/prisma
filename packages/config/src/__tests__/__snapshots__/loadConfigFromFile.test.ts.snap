// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadConfigFromFile invalid fails with \`TypeScriptImportFailed\` when the Prisma config file has a syntax error 1`] = `
"  [31mÃ—[0m Unexpected eof
   â•­â”€[[36;1;4m<prisma-config>.ts[0m:5:3]
 [2m3[0m â”‚ export default defineConfig({
 [2m4[0m â”‚   earlyAccess: true,
 [2m5[0m â”‚ }
   â•°â”€â”€â”€â”€


Caused by:
    Syntax Error"
`;
